Remove some unnecessary python functional wrappers

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #71993
* #61646
* __->__ #61608
* #71426

See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

peterbell10 committed:
Remove some unnecessary python functional wrappers

See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -771,7 +771,7 @@
   variants: function, method
 
-- func: bilinear(Tensor input1, Tensor input2, Tensor weight, Tensor? bias) -> Tensor
+- func: bilinear(Tensor input1, Tensor input2, Tensor weight, Tensor? bias=None) -> Tensor
```
cc @mruberry is changing this signature BC?

mruberry replied:
I don't think so, no - I think this is an FC issue, however, do you think that's an issue here? Will people begin calling the signature without bias and then old version of PyTorch won't know how to deal with those calls?

mruberry replied:
We're actually having a discussion about this now -- let me follow-up with those teams

mruberry replied:
OK so the way the jit works is that default arguments are not serialized unless they're explicitly specified. So I think this is an FC issue. We could verify by serializing a model with bilinear (without bias called) when PyTorch is built with this commit and then try loading it on PyTorch build before this commit.

peterbell10 replied:
@mruberry If I split the `=None` change into it's own PR and that gets landed first, would that be okay since the python wrapper will explicitly pass the bias argument?

mruberry replied:
If all the bilinear callsites pass bias explicitly then I think this will be OK from an FC perspective

albanD reviewed (commented):
The torch and nn functional changes look good.
I think it would be better if someone familiar with jit/fx reviews the rest.

ezyang commented:
I'm good for this. I bet the test changes are benign. @albanD you gonna land this?

albanD commented:
I would like someone from jit/fx to make sure that that part is ok. But if you think the whole thing is good, I can do the land, sure

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

mruberry commented:
FYI:

while torch.functional.istft does not appear to be used within Facebook it is used (rarely) in OSS and we should probably deprecate it.

ezyang commented:
> I would like someone more familiar with jit/fx to make sure that that part is ok. But if you think the whole thing is good, I can do the land, sure

@albanD OK, then you should find reviewers for the other parts. I see a few parts that should have FYIs anyway; the loss of ONNX input naming should go to the ONNX peeps, `test_swap_functional_linear` needs to go to quantization team since it's just a straight up bug (file follow up issue), and yes, someone from FX should give a quick onceover. You're better placed than @peterbell10 to identify the correct reviewers here and round them up.

albanD commented:
cc @eellison and @jamesr66a could you take a look at the jit/fx parts of this PR please?

cc @SplitInfinity do you know someone from the onnx team that can take a quick look at this please?

jamesr66a reviewed (commented):
FX parts LGTM

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

[ghstack-poisoned]

ezyang commented:
@SplitInfinity  is not here anymore, cc @pytorch/onnx-export let's also try @BowenBao

BowenBao commented:
> @SplitInfinity is not here anymore, cc @pytorch/onnx-export let's also try @BowenBao

onnx change looks good.

albanD reviewed (approved):
ok then

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31401361).

albanD commented:
This actually break some jit related internal tests. It will take a bit to land until I have time to check what happens.

peterbell10 committed:
Rebase and fix merge conflicts on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

mruberry commented:
Ping reminder for this pr @peterbell10 and @albanD -- per offline discussion we want to close some of these PRs!

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31401361).

albanD commented:
The PR below this one was added afterwards though? Should that be published/reviewed?

peterbell10 commented:
Yes, this has picked up a non-trivial merge conflict that needs be resolved. There is new fx-based quantization code that relies on the faulty assumption that certain arguments will always be passed as either positional or keyword arguments and this PR breaks that assumption because C++ functions pass through arguments to `__torch_function__` exactly as the function was called.

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

I've kept the symbol `torch.functional.istft` because it looks like public API,
but it could just as easily be moved to `_torch_docs.py`.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31401361).

albanD commented on torch/functional.py:
```diff
@@ -695,80 +696,74 @@ def stft(input: Tensor, n_fft: int, hop_length: Optional[int] = None,
-                     normalized, onesided, length, return_complex)
+istft = _add_docstr(
+    torch.istft,
+    "istft(input, n_fft, hop_length=None, win_length=None, window=None, center=True, "
```
Actually, the c++ function calls the first argument "self" instead of input.
This means that calling `torch.functional.istft(input=foo, ...)` used to work but won't anymore after this change.
That being said, today, you have to do `torch.istft(self=, ...)` if you use the torch. version...

Note that torchaudio does call the function with input= here: https://github.com/pytorch/audio/blob/ed6256a26aef254ecd048d795dc5814a1f570851/torchaudio/functional/functional.py#L177

Not sure what to do here as there might be uses for the torch. version with self= as well in the wild...

Can you remove that change from this PR for now so that we can land the PR and discuss this more problematic change in a different one?

peterbell10 replied:
The `self` argument in `native_functions.yaml` is renamed to `input` for the functional variants. You can see this in `python_torch_functionsEverything.cpp`:
```cpp
// istft
static PyObject * THPVariable_istft(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "istft(Tensor input, int64_t n_fft, int64_t? hop_length=None, int64_t? win_length=None, Tensor? window=None, bool center=True, bool normalized=False, bool? onesided=None, int64_t? length=None, bool return_complex=False)",
  }, /*traceable=*/true);

  // ...
}
```
and if I try to pass `self=foo` then it fails:
```
----> 1 torch.functional.istft(self=torch.ones(()), n_fft=2)
TypeError: istft() missing 2 required positional argument: "input", "n_fft"
```

albanD replied:
But torchscript doesn't know about that :( 
So when torchaudio tries to torchscript their model it blows up:
```
Argument self not provided.
:
  File "...../pytorch/audio/test/torchaudio_unittest/transforms_torchscript_consistency_cpu_test#binary,link-tree/torchaudio/functional/functional.py", line 176

    # default values are consistent with librosa.core.spectrum._spectrogram
    waveform = torch.istft(
               ~~~~~~~~~~~ <--- HERE
        input=spectrogram,
        n_fft=n_fft,
'inverse_spectrogram' is being compiled since it was called from 'InverseSpectrogram.forward'
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/pytorch/audio/test/torchaudio_unittest/transforms_torchscript_consistency_cpu_test#binary,link-tree/torchaudio/transforms.py", line 205
            Tensor: Dimension (..., time), Least squares estimation of the original signal.
        """
        return F.inverse_spectrogram(
               ~~~~~~~~~~~~~~~~~~~~~~
            spectrogram,
            ~~~~~~~~~~~~
            length,
            ~~~~~~~
            self.pad,
            ~~~~~~~~~
            self.window,
            ~~~~~~~~~~~~
            self.n_fft,
            ~~~~~~~~~~~
            self.hop_length,
            ~~~~~~~~~~~~~~~~
            self.win_length,
            ~~~~~~~~~~~~~~~~
            self.normalized,
            ~~~~~~~~~~~~~~~~
            self.center,
            ~~~~~~~~~~~~
            self.pad_mode,
            ~~~~~~~~~~~~~~
            self.onesided,
            ~~~~~~~~~~~~~ <--- HERE
        )

```

albanD replied:
cc @eellison is there any plan to be able to fix these? At the moment since script expects self= and the python binding expects input= we cannot use kwargs here?

peterbell10 replied:
Okay, I've removed the istft changes from this PR and opened an issue for the TorchScript bug #71994

albanD replied:
Awesome! Thanks!

peterbell10 committed:
Update on "Remove some unnecessary python functional wrappers"


See #61544 for an example of issues created by functional wrappers. In this
case, these are directly wrapping the native function with no added
functionality. One exception was `bilinear` which was just missing the default
argument in C++, but was otherwise the same.

Differential Revision: [D31401361](https://our.internmc.facebook.com/intern/diff/D31401361)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31401361).
